index.bs

```diff
@@ -9,8 +9,184 @@ Editor: Kent Tamura, Google https://www.google.com/, tkent@google.com
+  }
+}
+
+customElements.define('labeled-checkbox', LabeledCheckbox);
+&lt;/script>
+
+&lt;style>
+labeled-checkbox { border: dashed red; }
+labeled-checkbox:state(checked) { border: solid; }
+&lt;/style>
+
+&lt;labeled-checkbox>You need to check this&lt;/labeled-checkbox>
+
+<!-- Works even on ::part()s -->
+&lt;script>
+class QuestionBox extends HTMLElement {
+  constructor() {
+    super();
+    const shadowRoot = this.attachShadow({mode: 'closed'});
+    shadowRoot.innerHTML =
+      &#96;&lt;div>&lt;slot>Question&lt;/slot>&lt;/div>
+       &lt;labeled-checkbox part='checkbox'>Yes&lt;/labeled-checkbox>&#96;;
+  }
+}
+customElements.define('question-box', QuestionBox);
+&lt;/script>
+
+&lt;style>
+question-box::part(checkbox) { color: red; }
+question-box::part(checkbox):state(checked) { color: green; }
+&lt;/style>
+
+&lt;question-box>Continue?&lt;/question-box>
+&lt;/body>
+</pre>
+</div>
+
+
+Exposing custom element states {#exposing}
+============================
+
+Each <a>autonomous custom element</a> has <dfn>custom states</dfn>, a [=/set=]
+of strings. It's an empty set initially.
+
+<pre class=idl>
+partial interface ElementInternals {
+  [SameObject, PutForwards=value] readonly attribute DOMTokenList states;
+};
+</pre>
+
+The {{states}} IDL attribute returns a [=non-attribute DOMTokenList=] object
+associated with context object's
+<a href="http://html.spec.whatwg.org/C/#internals-target">target element</a>
+and its [=custom states=].
+
+<div class="example" id="ex-non-boolean-state">
+[=Custom states=] can expose boolean states represented by
+existence/non-exsitence of string tokens. If an author wants to expose a state
+which can have three values, it can be converted to three boolean states.
+For example, a state called <code>readyState</code> with <code>"loading"</code>,
+<code>"interactive"</code>, and <code>"complete"</code> values can be mapped to
+three boolean states, <code>"readyState_loading"</code>,
+<code>"readyState_interactive"</code>, and <code>"readyState_complete"</code>.
+
+<pre class="lang-js">
+// Change the readyState from anything to "complete".
+this.readyState = "complete";
+this.internals.<l>{{states}}</l>.<l>{{DOMTokenList/remove}}</l>("readyState_loading", "readyState_interactive");
+this.internals.<l>{{states}}</l>.<l>{{DOMTokenList/add}}</l>("readyState_complete");
+</pre>
+</div>
+
+## Non-attribute DOMTokenList ## {#sec-non-attribute-domtokenlist}
+
+This section defines a variant of {{DOMTokenList}}, called
+<dfn>non-attribute DOMTokenList</dfn>. It has the following differences from
+{{DOMTokenList}}.
+
+- A [=non-attribute DOMTokenList=] has an associated element and an associated
+    <a>token set</a>, but no attribute local name.
+
+- A [=non-attribute DOMTokenList=] object's
+    <a href="https://dom.spec.whatwg.org/#concept-dtl-update">update steps</a> are:
+
+    1. Recompute the style of the associated element.
+      <!-- TODO: How to represent this by spec terms? -->
```

Comment:
I think probably from a spec perspective its update steps are to do nothing. Recomputing the style is an unobservable implementation strategy, I think? Instead we define when :state() matches, as below. I assume the universe of CSS specs has some kind of existing statement that when the matching style rules change, the rendering on the screen must change. So just by defining when :state() matches we should get that for free.

Reply:
I checked several CSS-related specifications about this, and found nothing.  So, I changed this to "A non-attribute DOMTokenList has no update steps".